torch/csrc/jit/passes/common_subexpression_elimination.cpp

```diff
@@ -0,0 +1,145 @@
+#include "torch/csrc/jit/ir.h"
+
+#include <algorithm>
+#include <unordered_map>
+
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/jit/passes/common_subexpression_elimination.h"
+
+namespace torch { namespace jit {
+
+
+
+// Check whether two nodes have the same attributes in CSE.
+// This function may be too conservative for general use.
+// Do NOT support t/ts/g/gs attributes.
+// If t/ts are supported, CONSTANT node comparison may need to consider device.
+bool attributesEqualCSE(const Node* lhs, const Node* rhs) {
+  JIT_ASSERT(lhs != nullptr);
+  JIT_ASSERT(rhs != nullptr);
+  // One has attributes, the other does not.
+  if (lhs->hasAttributes() != rhs->hasAttributes()) return false;
+  // Neither has attributes.
+  if (!lhs->hasAttributes() && !rhs->hasAttributes()) return true;
+
+  auto lnames = lhs->attributeNames();
+  auto rnames = rhs->attributeNames();
+  if (lnames != rnames) return false;
+
+  Node* l = const_cast<Node*>(lhs);
+  Node* r = const_cast<Node*>(rhs);
+  for (auto name : lnames) {
+    if (l->kindOf(name) != r->kindOf(name)) return false;
+
+    #define COMPARE_ATTRIBUTEVALUE(type) \
+      case AttributeKind::type: \
+        { if (l->type(name) != r->type(name)) return false; } break;
+
+    switch(l->kindOf(name)) {
+      COMPARE_ATTRIBUTEVALUE(f)
+      COMPARE_ATTRIBUTEVALUE(fs)
+      COMPARE_ATTRIBUTEVALUE(i)
+      COMPARE_ATTRIBUTEVALUE(is)
+      COMPARE_ATTRIBUTEVALUE(s)
+      COMPARE_ATTRIBUTEVALUE(ss)
```

Comment:
NB: Nodes with graphs will silently compare False. This is OK for now but something to keep in mind.

Reply:
Add a comment.